Add: Challenge 완료 처리 및 메달 수여, Challenge 정보 수정 및 삭제, 포기 기능

won0845 opened the pull request:
## #️⃣ Part

- [ ] FE
- [x] BE

## 📝 작업 내용

>
### 1. 프론트에서 score API 호출 시 하루 챌린지 성공했다고 판단하고 쿠키 생성
### 2. challenge Entity 수정 
### 3. ChallengeController, ChallengeService, UserService 수정

---
#### Controller
1. 챌린지 하드 삭제
2. **챌린지 만료시 delete 및 complete 반환값 String 인데 예외처리 및 반환값 수정 필요**

#### Service
1. CreateChallenge
   a. endDate 추가 및 expired, deleted 컬럼 추가
   b. save 시 기존 인자를 그대로 save하던 에러 newChallenge로 수정 변경
2. editChallenge 기능 추가 -> (챌린지 시작 전) 시작날짜, 기상시간, 기간 3가지 변경 가능
3. deleteChallenge 구현 -> (챌린지 시작 전, host 권한만) challengeId와 hostId를 받아서 있는 지 확인 후 하드 삭제
4. ChallengeGiveUP 구현
   a. 챌린지 저장 및 user challege정보 -1 처리
   b. 호출자가 호스트인 경우 다른 인원에게 호스트 넘김
   c. 호출자가 challenge의 마지막 참여자인 경우 delete 처리
5. endChallenge 구현 (만료 처리 활용 목적) -> 날짜 비교해서 true, false 반환 -> **시간까지 비교필요**
6. complete 챌린지 (챌린지 만료 처리) 구현
   a. 챌린지 시간 비교해서 아직 끝나지 않은경우 return
   b. user의 챌린지 정보를 -1 처리
   c. 호스트인지 체크 후 expired로 변경
   d. 메달 수여
      -  기간별로 90%이상 80점 이상 달성시 메달 획득 금 100 은 30 동 7

#### userService
1. resetChallenge 구현 -> userID를 받아서 -1 처리 후 user정보 redis cache 삭제
2. updateUserMedals 구현 -> userId와 메달 타입을 입력받음 -> 메달수 증가 후 저장
3. decideMedalType 구현 -> 기간을 입력받아서 메달 type 반환

won0845 committed:
feat : 프론트에서 점수 API 호출시 게임 종료 되었다고 판단하여 오늘 하루쿠키 남김

won0845 committed:
feat : challenge Entity 변경

end_Date, expired, deleted 컬럼 추가

won0845 committed:
feat : 챌린지 삭제, 챌린지 종료(만료)시

Controller
1.  챌린지 하드 삭제
2. 챌린지 만료시

delete 및 complete 반환값 String 인데 예외처리 및 반환값 수정 필요

Service
1.CreateChallenge
   a.endDate추가 및 expired, deleted 컬럼 추가
   b. save시 인자를 save하는 에러를 newChallenge로 변경
2. editChallenge 추가-> 시작날짜, 기상시간, 기간 3가지 변경가능
3. endChallenge 구현 -> 날짜 비교해서 true, false 반환 -> 시간까지 비교필요

4. deleteChallenge 구현 -> challengeId와 hostId를 받아서 있는 지 확인 후 하드 삭제

5. ChallengeGiveUP 구현 ->
   a. 마지막인경우 delete 처리
   b. 호스트 인경우 다른 인원에게 호스트 넘김
   c. 챌린지 저장 및 user challege정보 -1 처리
6. complete 챌린지 구현
   a. 챌린지 시간 비교해서 아직 끝나지 않은경우 return
   b. user챌린지 정보를 -1 처리
   c. 호스트인지 체크 후 expired로 변경
   d. 메달처리
      -  기간별로 90%이상 80점 이상 달성시 메달 획득 금 100 은 30 동 7

userService
1. resetChallenge 구현 -> userID를 받아서 -1 처리 후 user정보 redis cache 삭제
2. updateUserMedals 구현 -> userId와 메달 타입을 입력받음 -> 메달수 증가 후 저장
3. decideMedalType 구현 -> 기간을 입력받아서 메달 type 반환

won0845 committed:
feat : challenge 정보 조회 시 hostID 추가

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +41,111 @@ export class ChallengesService {
+    editChall.deleted = false;
+
+    return await this.challengeRepository.save(editChall);
+  }
```
front에서 방어를 하고는 있지만, 조작하면 얼마든 할 수 있는 브라우저 특성을 고려해서..
아래와 같은 검증 함수가 추가되면 좋겠습니다.

1. startDate + wakeTime이 현재시점보다 이후인지 검사하는 메소드,
2. duration이 7, 30, 100 중 하나인지 확인하는 메소드

그래서 위의 검사에 모두 통과한 경우에만 create과 edit이 가능하도록 처리하고, 아닌 경우 각각에 해당하는 error를 뱉는 것으로요!

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +41,111 @@ export class ChallengesService {
+
+  // 챌린지 ID 랑 userID 둘다 받아서 호스트인지 확인하고 삭제로 수정
+  // 호스트ID 하나로만 조회 ?? -> 챌린지ID랑 호스트 ID 말고 ?
+  // 모두가 호출가능 ?? -> 프론트 또는 백에서 host인지 아닌지 확인 후 그에따른 결과값 송출 ???
```
지난번에 온라인 미팅에서 논의했던 것처럼, front에서 user가 host인지 확인할 것이지만, back에서도 이중으로 검사하면 좋겠습니다.

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +395,40 @@ export class ChallengesService {
+  async completeChallenge(challengeId: number, userId: number): Promise<void> {
+    if (!this.endChallenge(challengeId)) {
+      // 챌린지 시간 비교해서 아직 끝나지 않았다면 return 처리 해야함
+      // -> error를 발생시켜야 하나?
```
아직 challenge가 만료되지 않은 경우 {expired: false, message: 'This challenge is not expired yet'} 이런식의 값을 return하고, 만료된 경우, 모든 일을 다 처리한 뒤, {expired: true, message: 'This challenge has expired and the user data has been successfully updated'}를 return 하면 어떤가 하는 생각이 듭니다. 이 API를 호출했을때 front 측에서 expired 상태를 일관성있게 확인할 수 있도록요~

pkw-May commented on backend/src/challenges/challenges.entity.ts:
```diff
@@ -32,6 +32,15 @@ export class Challenges {
+  @Column({ name: 'end_date' })
+  endDate: Date;
+
+  @Column({ name: 'expired' })
+  expired: boolean;
```
우리가 처음 논의할 때는 expired column을 넣자고 얘기하긴 했지만... 로직을 쭉 보면은 expired라는 column의 필요성이 크게 없는 거 같다는 생각이 드는데 어떤가요?

pkw-May replied:
추후 데이터 활용 가능성을 염두하여 deleted와 함께 컬럼 유지하기로 결정.

두 column이 존재하면 완료되어서 끝난 것인지, 완료 전에 멈춘 것인지 확인 등이 가능함

pkw-May commented on backend/src/challenges/challenges.entity.ts:
```diff
@@ -32,6 +32,15 @@ export class Challenges {
+  @Column({ name: 'expired' })
+  expired: boolean;
+
+  @Column({ name: 'deleted' })
+  deleted: boolean;
```
deleted column도... 우리가 hard delete하자고 결정을 했으니, 굳이 column이 존재하지 않아도 될거 같아요

pkw-May reviewed (commented):
고생 많으셨습니다!! 온라인 회의에서 말씀하셨던 '자잘하게 손봐야 할 것들' 업데이트 하시면서, 제가 남긴 리뷰도 한 번 확인해서 반영 부탁드립니다~

won0845 committed:
feat : Edit Challenge

1. editChallenge controller추가
2. EditChallengeDTO 추가
3. challengeGiveUp if 문 수정

won0845 committed:
feat : Edit endChallenge

1. endChallenge 수정 -> 날짜만 비교 -> 시간까지 같이 비교한다.

캐시삭제는 다른 팀원들도 남아있을 수 있기에 삭제하지 않기로

2. completeChallenge에서 호스트 인경우 expired 하고 저장할 때 redis에서 삭제

won0845 committed:
feat : Edit updateUserMedals

1. 메달 수 수정하고 저장 시 redis에서 유저 정보 삭제

won0845 committed:
feat : Edit submitScore

1. 챌린지 만료 됬는지 여부 로직에서 게임 종료시 확인하고 홈에서 한번 더 확인하는 방법으로 수정

-> 게임 종료시 서버에 점수 저장 후 메달처리까지 하면서 사용자 경험 상승 및 홈에서 한번 더 확인으로 이중 확인

-> 홈에서는 로컬에 저장된 챌린지 정보를 비교해서 현재 시간과 챌린지 마지막 끝나는 기상시간이 지난 경우 api 호출

won0845 committed:
feat : edit delete and complete

1. controller에서 delete및 complete시 문자열 반환에서 challenge 반환으로 변경

won0845 committed:
feat : create 및 edit 시 검증 메소드 추가

1.startDate + wakeTime이 현재시점보다 이후인지 검사

2. duration이 7, 30, 100 중 하나인지 검사

3. edit 후 challenge 캐시 삭제

won0845 committed:
feat : challenge 및 user 정보 변경 시 캐시 삭제 추가

won0845 committed:
feat : challenge가 만료된 경우 만료되지 않은 경우 response 추가

1. 아직 challenge가 만료되지 않은 경우 {expired: false, message: 'This challenge is not expired yet'} 리턴
2. 챌린지 만료된 경우 {expired: true, message: 'This challenge has expired and the user data has been successfully updated'}를 리턴

won0845 committed:
Merge branch 'dev-v2' into backend/challengeComplete

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +43,119 @@ export class ChallengesService {
+    if (!editChall) {
+      //host 권한이 없는 챌린지 수정 시 에러처리
+      throw new NotFoundException(
+        `Challenge with ID ${challenge.hostId} not found`,
```
front에서 body에 담아보낸 hostId값이 challenge의 hostId와 일치하지 않는 것이므로, 에러메시지는 "User with ID ${challenge.hostId} is not the host of this challenge"라고 보내주시는 게 더 명확할 거 같습니다.

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +43,119 @@ export class ChallengesService {
+      where: { _id: challengeId, hostId: hostId },
+    });
+    if (!challenge) {
+      throw new NotFoundException(`Challenge with ID ${challengeId} not found`);
```
여기도 edit의 경우와 마찬가지로, delete 권한이 없는 유저가 삭제를 요청한 상황이므로 에러메시지는 "User with ID ${challenge.hostId} is not the host of this challenge"라고 쓰는 것이 명확하겠습니다.

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +43,119 @@ export class ChallengesService {
+    return challenge;
+  }
+
+  async challengeGiveUp(challengeId: number, userId: number): Promise<void> {
```
현재 코드상에서 challengeGiveUp service에 대응되는 controller가 생성되지 않은것으로 보입니다. 확인 부탁드려요.

pkw-May replied:
또, GiveUp하는 challenge는 이미 challenge가 진행중인 것에 대해서만 가능합니다. challenge 시작 전이나, 완료 후에 API가 요청될일은 없도록 front에서 방어하겠지만,, 혹시 모르니 back에서도 해당 API의 요청 시점이 challenge 진행중인지 확인하는 validate 메소드를 하나 추가하면 좋을 듯 합니다.

pkw-May commented on backend/src/challenges/challenges.controller.ts:
```diff
@@ -58,6 +58,50 @@ export class ChallengesController {
+  }
+
+  // 로컬에 저장한 챌린지 값으로 현재 날짜랑 챌린지 날짜 비교해서 넘은 경우만 호출
+  @Post('complete/:challengeId/:userId') // 챌린지가 끝났는지 확인하는 경우
```
column 명이 expired이니까 일관성을 위해서, API url과 service 등 메소드 이름도 모두 expired로 바꿉시다

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +43,119 @@ export class ChallengesService {
+    });
+    if (users.length === 1) {
+      // 혼자인경우 당연히 호스트인데 예외처리 해줘야하나?
+      challenge.deleted = true;
```
deleted 처리는 challenge 시작전에만 처리하는 속성이라서, giveUp의 경우 deleted column을 손대면 안 될것으로 판단됨. 예를들어 이미 며칠 진행한 challenge의 경우, deleted 처리해버리면, 며칠간의 기록마저 challenge history에서 지워져버리는 현상이 발생할듯. 중도포기의 경우 challenge history에서 진행한 만큼은 기록을 조회할 수 있도록, deleted가 아닌 다른 방식으로 data에 표기해야 할듯. gaveUp은 각 멤버마다 column을 따로 갖고 있어야 할듯...... model 자체가 달라져야 할거 같네.. 이거 다시한번 고민해보기ㅠ

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +406,73 @@ export class ChallengesService {
   }
+
+  // 날짜 비교해서 챌린지 끝난경우 호출되는 메소드
+  async completeChallenge(
```
이것도 controller에 언급했던 것처럼 일관성을 위해 complete 가 아닌 expired로 통일하는 것이 좋을듯합니다~

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +406,73 @@ export class ChallengesService {
+    if (!challenge) {
+      throw new NotFoundException(`Challenge with ID ${challengeId} not found`);
+    }
+    // 1. 호스트인지 체크 후 호스트 인경우 챌린지 expired로 변경 -> 챌린지 정보를 가져와야 알 수 있음 userID 랑 비교
```
왜 host인지가 중요한 것일까요..? challenge의 만료여부는 모든 멤버에게 공통된 데이터인것으로 생각이 되어요.. 흠... 논의가 필요해보입니다...!

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +406,73 @@ export class ChallengesService {
+
+    // 3. 메달처리
+    // 기간별로 90%이상 80점 이상 달성시 메달 획득 금 100 은 30 동 7
+    const cutLine = await this.attendanceRepository.find({
```
조건에 맞는 상세 데이터가 필요하기보다는 length만 알면 되니까, find가 아닌 count 메소드를 쓰는 것이 효율적이겠습니다. 그리고 상황상 이 count 변수는 user의 성적이라고 할 수 있으니, 기준이라고 이해되는 cutLine보다는 80점을 넘긴 날의 개수인 qualifiedDaysCount 정도로 표현하면 어떨까합니다

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +406,73 @@ export class ChallengesService {
+    const cutLine = await this.attendanceRepository.find({
+      where: { challengeId, userId, score: MoreThan(80) },
+    });
+    const total = challenge.duration;
```
사실상 여기서 total로 쓰인 변수값이 메달 수여 여부를 결정하니... 앞선 cutLine의 역할을 하는 것으로 보입니다. 근데 cutLine보다 자주쓰이는 표현인 threshold를 쓰면 어떨까 합니다..!

pkw-May reviewed (commented):
짧은시간동안 많은 기능을 추가하시느라 고생많으셨습니다. 추가된 코드에 대해 몇가지 수정 보완 의견을 덧붙였습니다..! 확인 부탁드려요

pkw-May commented:
재원님~ 제가 혼자 이런 저런 생각을 하면서
이 챌린지의 상태를 어떻게 관리를 할지 슥슥 그려본 게 있어요

![83B41EC4-81D0-4EBD-A8C4-19F2FF8BC781_1_105_c](https://github.com/user-attachments/assets/98d1c486-6328-4eb7-8f94-0ea1aea6d4cf)

참고 한번 해봐주세요~ 일요일에 의견 많이 내주시길 부탁드리며.. 이만..

won0845 committed:
Merge branch 'backend/challengeComplete' into backend/Edit_challengeComplete

won0845 committed:
feat : expired -> completed로 변경

won0845 committed:
feat: 예외처리 에러메세지 수정

won0845 committed:
feat: Edit completeChallenge

1. find 함수 count로 수정
2. cutLine 변수명 qualifiedDaysCount 로 변경
3. total 변수명threshold로 변경

won0845 committed:
feat: add challengeGiveUpController

1. 챌린지 기브업 컨트롤러 추가
2. challengeGiveUp service 예외처리 추가

won0845 committed:
feat: add validateChallengeDate

1. 챌린지 정보 바로 db가 아닌 redis에서 한번 확인 후 가져옴
2. challengeGiveUp시 챌린지가 현재 날짜 기준으로 진행중인지 체크

pkw-May commented on backend/src/challenges/challenges.controller.ts:
```diff
@@ -58,6 +58,50 @@ export class ChallengesController {
+    );
+    if (result === true) {
+      return {
+        expired: true,
```
DB column에서 completed 를 쓰기로 했으니 return 값도 completed를 쓰는게 맞을 거 같아요..

근데 이번에 또 이 변수명을 다시 보니깐 completed는 챌린지를 꾸준히 잘 참여해서 성공한, mission success의 느낌이 나는데 유저 중에는 그렇지 않은 경우도 많을 거라.. expired가 낫나? 하는 생각이 문득 드네요....ㅋ...... 치명적인 부분은 전혀 아닌데 그런 생각이 들어서.. 무튼 네.. 고려해봐주세요ㅎㅎㅎ

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +43,152 @@ export class ChallengesService {
+    return await this.challengeRepository.save(editChall);
+  }
+
+  async endChallenge(challengeId: number): Promise<boolean> {
```
혹시 함수의 위치를 여기에 두신 특별한 이유가 있을까요? 이 친구를 부르는 completeChallenge 함수가 저 밑에 있어서.. 가깝게 붙여놓으면 어떨까 해서요 ㅎㅎ...

그리고 이 함수는 요청날짜가 챌린지 종료 시점보다 늦는지 확인만 하고 데이터는 손대지 않으니까 함수명을 checkChallengeExpiration 같은 것으로 바꾸면 어떨까요?

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +460,90 @@ export class ChallengesService {
+    userId: number,
+  ): Promise<boolean> {
+    if (!this.endChallenge(challengeId)) {
+      // -> error를 발생시켜야 하나?
```
endChallenge 함수 내부에 에러 뱉는 구문이 있는데, try catch로 안 받아도 되나요..?  Controller에서 알아서 처리하려나.. 만일 그런경우 endChallenge때문에 발생한 에러라는 정보를 디버깅할 수 있는 구조인가요?

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -293,4 +460,90 @@ export class ChallengesService {
+      throw new NotFoundException(`Challenge with ID ${challengeId} not found`);
+    }
+    // 1. 호스트인지 체크 후 호스트 인경우 챌린지 expired로 변경 -> 챌린지 정보를 가져와야 알 수 있음 userID 랑 비교
+    if (challenge.hostId === userId) {
```
문득 든 생각인데, 만약 호스트씩이나 되어놓고도 챌린지 끝나고 영영 돌아오지 않는다면, 이 챌린지는 영영 completed가 업데이트 되지 않아버리는 문제가 발생할 수 있을 거 같아요.. 당장 급하거나 치명적인 문제는 아니지만.. 누구든 먼저 들어온 사람으로하여금 completed를 업뎃하게 하고, 이미 변경되어있다면 그대로 두는 방식으로 처리하는 건 어떨까요..? 물론 모두다 안들어와서 영영 구천을 떠도는 챌린지가 될 수도 있음...

pkw-May commented on backend/src/challenges/challenges.service.ts:
```diff
@@ -34,17 +43,152 @@ export class ChallengesService {
+    challengeId: number,
+    hostId: number,
+  ): Promise<Challenges> {
+    const challenge = await this.challengeRepository.findOne({
```
이거... 혹쉬 모르니깐 정말 챌린지 시작하기 전에 삭제 요청한 거 맞는지 체크하는 if문 하나 넣어주실 수 있을까요? challenge startDate가 오늘보다 이후인지 확인해서, 이전일때만 delete할 수 있도록

pkw-May reviewed (commented):
먼곳에서 고생많으셨습니다..! 몇가지 궁금한 점이 있어서 리뷰 달아놓았으니 확인부탁드려요~

pkw-May commented on backend/src/challenges/challenges.controller.ts:
```diff
@@ -58,6 +58,50 @@ export class ChallengesController {
+
+  // 로컬에 저장한 챌린지 값으로 현재 날짜랑 챌린지 날짜 비교해서 넘은 경우만 호출
+  @Post('complete/:challengeId/:userId') // 챌린지가 끝났는지 확인하는 경우
+  async checkChallenge(
```
얘 이름도 서비스 이름이랑 같이 맞추는 거 어떤가요? completeChallenge로요... 단순 챌린지 만료 확인은 내부 로직 중 하나니깐.. 목적자체는 만료처리이고 ㅇㅇ

won0845 committed:
feat : edit endChallenge

1. 함수 이름 변경 및 challenge 정보 redis에 있는 지 먼저 확인 후 db 조회

won0845 committed:
Merge branch 'dev-v2' into backend/challengeComplete

won0845 committed:
feat: change variable name from 'expired' to 'completed'

won0845 committed:
feat: change function name from isEndChallenge to checkChallengeExpiration

won0845 committed:
feat : change function name from checkChallenge to completeChallenge

won0845 committed:
feat : update function completeChallenge condition to avoid multiple updates

won0845 committed:
feat : deleteChallenge 반환값 수정

won0845 committed:
feat: completeChallenge 이중 DB호출 해결

won0845 committed:
feat : deleteChallenge 예외세분화 및 삭제 기준날짜 유효성 검사 추가

won0845 committed:
feat : errorCode 수정

won0845 committed:
fix : edit ChallengeService

1. checkChallengeExpiration에서 boolean값 객체로 감싸져있어 if 통과 못하는거 해결
2. wakeTime Date로 되어있지만 실질적으로 string이어서 getMethod()사용불가 string으로 변환 후 추출하여 해결
3. qualifiedDaysCount 조회시 db값 못 불러오는거 해결 -> db는 primary기준으로 조회되는데 primaryGeneratedColumn삭제 후 해결
4. 점수 80점 초과에서 80이상으로 변경

won0845 committed:
feat : editChallenge시 challengeId 추가

1.권한이 없는 사용자가 챌린지 바꾸는지 체크

won0845 committed:
feat: 챌린지 생성시 중복 제거

1. 챌린지 생성시 같은 호스트가 챌린지 여러개 만들 수 있는 문제 예외처리

won0845 committed:
feat : 이미 완료된 챌린지 수정 시 예외처리

pkw-May reviewed (approved):
오랜기간.. 또 타지에서 수많은 질문에 답해주시고, 또 엄청난 리뷰들 모두 반영해주셔서 감사합니다! 정말 수고 많으셨습니다!
